pkg/addons/manifest.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2020 The KubeOne Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package addons
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+
+	"github.com/kubermatic/kubeone/pkg/state"
+
+	metav1unstructured "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	kyaml "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+func getManifests(s *state.State) error {
+	vars := map[string]interface{}{
+		"KubeOne": s.Cluster,
+	}
+	manifests, err := loadAddonsManifests(s.Cluster.Addons.Path, s.Verbose, vars)
+	if err != nil {
+		return err
+	}
+
+	rawManifests, err := ensureAddonsLabelsOnResources(manifests)
+	if err != nil {
+		return err
+	}
+
+	combinedManifests := combineManifests(rawManifests)
+	s.Configuration.AddFile("addons/addons.yaml", combinedManifests.String())
+
+	return nil
+}
+
+// loadAddonsManifests loads all YAML files from a given directory and runs the templating logic
+func loadAddonsManifests(addonsPath string, verbose bool, vars map[string]interface{}) ([]runtime.RawExtension, error) {
+	manifests := []runtime.RawExtension{}
+
+	files, err := ioutil.ReadDir(addonsPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read the addons directory %s", addonsPath)
+	}
+
+	for _, file := range files {
+		filePath := filepath.Join(addonsPath, file.Name())
+		if file.IsDir() {
+			fmt.Printf("Found directory '%s' in the addons path. Ignoring.\n", file.Name())
+			continue
```

Comment:
I'm not really sure if we need to limit addons to flat structure of files

Reply:
I'm worried that it would just introduce unnecessary complexity. A big problem is in which order to apply manifests:
* should we first apply manifests from directories,
* should we first apply manifests from the root directory and then from other directories
* what about subdirectories (i.e. directory in the directory)

IMO this is way too complex and not worth it at all. With the flat structure of files, you can easily control the order just by setting the appropriate filename. This will not be that easy if we include directories too. Anyways, usually there is only one manifest per component since YAML allows multiple resources to be defined in a single manifest.

Reply:
OK, let's have it like this for now.